core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -209,47 +208,58 @@ public T call() throws Exception {
     private class JedisPool {
         private Queue<BinaryJedis> _queue = new ConcurrentLinkedQueue<BinaryJedis>();
 
-        public BinaryJedis borrowInstance(boolean knownGood) {
+        public BinaryJedis borrowInstance(boolean reinitializePool) {
+            if (reinitializePool) {
+                if (_log.isInfoEnabled())
+                    _log.info("Reinitializing Redis connection pool");
+                
+                // Shut down pool, effectively closing all connections
+                // Note that there is a chance that connections which are currently in use are still returned to the
+                // pool after the shutdown because of a race condition, but since we create a new instance below
+                // unconditionally we can at least be sure no operation will fail a second time
+                shutdown();
```

Comment:
There's also another race condition: 
* 2 redis requests R1 and R2 are in flight
* R1 fails and creates a new, healthy connection C for R3
* R3 completes and returns the healthy connection to the pool
* R2 fails
* This will also close the healthy connection

Most probably this is just inefficient, but in combination with the broad exception handling I fear that it might have worse consequences and might result in continuous connection creation / shutdown.

Reply:
I wouldn't see this as a prerequisite to merging this, as I'd assume that you'd submit a follow up PR if this causes issues :-)